File: tiktok/apis/display.py
"""Display API namespace.

Wraps the following TikTok endpoints:

* ``GET  /v2/user/info/``    — user profile information
* ``POST /v2/video/list/``   — paginated list of the user's videos
* ``POST /v2/video/query/``  — fetch specific videos by ID

Reference: https://developers.tiktok.com/doc/display-api-overview
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Any

from tiktok.apis.base import BaseAPI
from tiktok.config import TIKTOK_API_BASE_URL, VIDEO_LIST_MAX_COUNT, VIDEO_QUERY_MAX_IDS
from tiktok.exceptions import TikTokAPIError
from tiktok.models.display import (
    User,
    UserField,
    Video,
    VideoField,
    VideoListData,
    VideoQueryData,
)

_BASE = TIKTOK_API_BASE_URL


class DisplayAPI(BaseAPI):
    """Methods for reading TikTok user and video data.

    Requires an access token authorised with at minimum
    ``user.info.basic`` and/or ``video.list`` scopes depending on the
    method being called.
    """

    # ------------------------------------------------------------------
    # User info
    # ------------------------------------------------------------------

    async def get_user_info(self, fields: list[UserField]) -> User:
        """Retrieve profile information for the authenticated user.

        Parameters
        ----------
        fields:
            One or more :class:`~tiktok.models.display.UserField` values to
            include in the response.  Only requested fields will be populated
            on the returned :class:`~tiktok.models.display.User` object.

        Scope required: ``user.info.basic`` (plus ``user.info.profile`` and/or
        ``user.info.stats`` for the corresponding fields).

        Returns:
            :class:`~tiktok.models.display.User` with the requested fields.

        Raises:
            :class:`~tiktok.exceptions.TikTokAuthError`: Token lacks scope.
            :class:`~tiktok.exceptions.TikTokAPIError`: Any other API error.
        """
        params: dict[str, Any] = {"fields": _join_fields(fields)}
        payload = await self._session.get(f"{_BASE}/v2/user/info/", params=params)
        return User.model_validate(payload["data"]["user"])

    # ------------------------------------------------------------------
    # Video list
    # ------------------------------------------------------------------

    async def list_videos(
        self,
        fields: list[VideoField],
        *,
        cursor: int | None = None,
        max_count: int = VIDEO_LIST_MAX_COUNT,
    ) -> VideoListData:
        """Return one page of the authenticated user's public videos.

        Videos are sorted by ``create_time`` in descending order (newest
        first).  Use :attr:`~tiktok.models.display.VideoListData.cursor`
        and :attr:`~tiktok.models.display.VideoListData.has_more` to
        paginate, or use :meth:`iter_videos` for an automatic async iterator.

        Parameters
        ----------
        fields:
            Video fields to include in each returned
            :class:`~tiktok.models.display.Video` object.
        cursor:
            Pagination cursor (UTC Unix timestamp in milliseconds) from the
            previous response.  Omit to start from the most recent video.
        max_count:
            Maximum number of videos per page (default: ``20``, max: ``20``).

        Scope required: ``video.list``

        Returns:
            :class:`~tiktok.models.display.VideoListData`.

        Raises:
            :class:`~tiktok.exceptions.TikTokAuthError`: Token lacks scope.
            :class:`~tiktok.exceptions.TikTokAPIError`: Any other API error.
        """
        params: dict[str, Any] = {"fields": _join_fields(fields)}
        body: dict[str, Any] = {"max_count": max_count}
        if cursor is not None:
            body["cursor"] = cursor

        payload = await self._session.post(f"{_BASE}/v2/video/list/", json=body, params=params)
        return VideoListData.model_validate(payload["data"])

    async def iter_videos(
        self,
        fields: list[VideoField],
        *,
        max_count: int = VIDEO_LIST_MAX_COUNT,
    ) -> AsyncIterator[Video]:
        """Async generator that yields every video for the authenticated user.

        Automatically follows pagination cursors until all videos have been
        yielded.

        Parameters
        ----------
        fields:
            Video fields to include in each yielded
            :class:`~tiktok.models.display.Video` object.
        max_count:
            Page size for each internal request (default: ``20``).

        Scope required: ``video.list``

        Yields:
            :class:`~tiktok.models.display.Video` instances.

        Example::

            async for video in client.display.iter_videos([VideoField.ID, VideoField.TITLE]):
                print(video.id, video.title)
        """
        cursor: int | None = None
        while True:
            page = await self.list_videos(fields, cursor=cursor, max_count=max_count)
            for video in page.videos:
                yield video
            if not page.has_more:
                break
            cursor = page.cursor

    # ------------------------------------------------------------------
    # Video query
    # ------------------------------------------------------------------

    async def query_videos(
        self,
        video_ids: list[str],
        fields: list[VideoField],
    ) -> VideoQueryData:
        """Fetch metadata for specific videos by their IDs.

        The authenticated user must own all requested videos.  Up to
        ``VIDEO_QUERY_MAX_IDS`` (20) IDs may be requested per call.

        Parameters
        ----------
        video_ids:
            List of video IDs to look up (1 - 20 items).
        fields:
            Video fields to include in the response.

        Scope required: ``video.list``

        Returns:
            :class:`~tiktok.models.display.VideoQueryData`.

        Raises:
            :class:`~tiktok.exceptions.TikTokAPIError`:
                If ``video_ids`` is empty or exceeds ``VIDEO_QUERY_MAX_IDS``.
        """
        if not video_ids:
            raise TikTokAPIError(
                code="invalid_params",
                message="video_ids must contain at least one ID.",
                log_id="",
            )
        if len(video_ids) > VIDEO_QUERY_MAX_IDS:
            raise TikTokAPIError(
                code="invalid_params",
                message=(
                    f"video_ids may contain at most {VIDEO_QUERY_MAX_IDS} IDs "
                    f"per request; {len(video_ids)} were provided."
                ),
                log_id="",
            )

        params: dict[str, Any] = {"fields": _join_fields(fields)}
        payload = await self._session.post(
            f"{_BASE}/v2/video/query/",
            json={"filters": {"video_ids": video_ids}},
            params=params,
        )
        return VideoQueryData.model_validate(payload["data"])


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _join_fields(fields: list[UserField] | list[VideoField]) -> str:
    """Serialise a list of field enums into a comma-separated string."""
    return ",".join(f.value for f in fields)

File: go/display.go
package tiktok

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

// User mirrors the subset of TikTok user fields exposed via /v2/user/info/.
// Fields are populated only if they were requested via the `fields` argument.
type User struct {
	OpenID          string `json:"open_id,omitempty"`
	UnionID         string `json:"union_id,omitempty"`
	AvatarURL       string `json:"avatar_url,omitempty"`
	AvatarURL100    string `json:"avatar_url_100,omitempty"`
	AvatarLargeURL  string `json:"avatar_large_url,omitempty"`
	DisplayName     string `json:"display_name,omitempty"`
	BioDescription  string `json:"bio_description,omitempty"`
	ProfileDeepLink string `json:"profile_deep_link,omitempty"`
	IsVerified      bool   `json:"is_verified,omitempty"`
	Username        string `json:"username,omitempty"`
	FollowerCount   int64  `json:"follower_count,omitempty"`
	FollowingCount  int64  `json:"following_count,omitempty"`
	LikesCount      int64  `json:"likes_count,omitempty"`
	VideoCount      int64  `json:"video_count,omitempty"`
}

// Video mirrors the TikTok video resource fields exposed via /v2/video/list/
// and /v2/video/query/.
type Video struct {
	ID                string   `json:"id,omitempty"`
	CreateTime        int64    `json:"create_time,omitempty"`
	CoverImageURL     string   `json:"cover_image_url,omitempty"`
	ShareURL          string   `json:"share_url,omitempty"`
	VideoDescription  string   `json:"video_description,omitempty"`
	Duration          int64    `json:"duration,omitempty"`
	Height            int64    `json:"height,omitempty"`
	Width             int64    `json:"width,omitempty"`
	Title             string   `json:"title,omitempty"`
	EmbedHTML         string   `json:"embed_html,omitempty"`
	EmbedLink         string   `json:"embed_link,omitempty"`
	LikeCount         int64    `json:"like_count,omitempty"`
	CommentCount      int64    `json:"comment_count,omitempty"`
	ShareCount        int64    `json:"share_count,omitempty"`
	ViewCount         int64    `json:"view_count,omitempty"`
	HashtagNames      []string `json:"hashtag_names,omitempty"`
	VideoLabel        any      `json:"video_label,omitempty"`
	IsPaidPartnership bool     `json:"is_paid_partnership,omitempty"`
}

// VideoListResult is the page wrapper returned by /v2/video/list/.
type VideoListResult struct {
	Videos  []Video `json:"videos"`
	Cursor  int64   `json:"cursor"`
	HasMore bool    `json:"has_more"`
}

// VideoQueryResult is the wrapper returned by /v2/video/query/.
type VideoQueryResult struct {
	Videos []Video `json:"videos"`
}

// GetUser fetches the authenticated user's profile. `fields` must contain at
// least one TikTok user field name (e.g. "open_id", "display_name",
// "follower_count").
//
// Scope: user.info.basic (and user.info.profile / user.info.stats for the
// corresponding fields).
func (c *Client) GetUser(ctx context.Context, fields []string) (*User, error) {
	if len(fields) == 0 {
		return nil, errors.New("tiktok: GetUser requires at least one field")
	}
	q := url.Values{}
	q.Set("fields", strings.Join(fields, ","))

	data, err := c.doJSON(ctx, http.MethodGet, c.baseURL+"/v2/user/info/", nil, q)
	if err != nil {
		return nil, err
	}
	var wrapper struct {
		User User `json:"user"`
	}
	if err := json.Unmarshal(data, &wrapper); err != nil {
		return nil, fmt.Errorf("tiktok: decode user: %w", err)
	}
	return &wrapper.User, nil
}

// ListVideosParams configures a single page request to /v2/video/list/.
type ListVideosParams struct {
	Fields   []string // Required. Video fields to request.
	Cursor   *int64   // Pagination cursor (UTC Unix ms). Nil = first page.
	MaxCount int      // Page size (1-20). Defaults to 20.
}

// ListVideos returns one page of the authenticated user's videos.
//
// Scope: video.list.
func (c *Client) ListVideos(ctx context.Context, p ListVideosParams) (*VideoListResult, error) {
	if len(p.Fields) == 0 {
		return nil, errors.New("tiktok: ListVideos requires at least one field")
	}
	maxCount := p.MaxCount
	if maxCount <= 0 {
		maxCount = 20
	}
	body := map[string]any{"max_count": maxCount}
	if p.Cursor != nil {
		body["cursor"] = *p.Cursor
	}

	q := url.Values{}
	q.Set("fields", strings.Join(p.Fields, ","))

	data, err := c.doJSON(ctx, http.MethodPost, c.baseURL+"/v2/video/list/", body, q)
	if err != nil {
		return nil, err
	}
	out := &VideoListResult{}
	if err := json.Unmarshal(data, out); err != nil {
		return nil, fmt.Errorf("tiktok: decode video list: %w", err)
	}
	return out, nil
}

// IterVideosParams configures IterVideos.
type IterVideosParams struct {
	// Fields is required: the video fields to request on each page.
	Fields []string
	// MaxCount is the per-page size (1-20). Defaults to 20 when non-positive.
	MaxCount int
	// Limit caps the total number of videos returned across all pages. When
	// non-positive, every available video is returned. It bounds memory and
	// upstream calls so the iteration can never run unbounded.
	Limit int
}

// IterVideos returns every video for the authenticated user, automatically
// following pagination cursors until all videos have been collected (or the
// optional Limit is reached). It is the Go equivalent of the Python SDK's
// iter_videos async generator; because Go has no async generators, the SDK
// follows the established cursor-walking convention (see GetAccountAnalytics)
// and returns the accumulated slice.
//
// Each page request honours ctx; cancellation stops the walk and surfaces the
// context error.
//
// Scope: video.list.
func (c *Client) IterVideos(ctx context.Context, p IterVideosParams) ([]Video, error) {
	if len(p.Fields) == 0 {
		return nil, errors.New("tiktok: IterVideos requires at least one field")
	}
	maxCount := p.MaxCount
	if maxCount <= 0 {
		maxCount = 20
	}

	var (
		all    []Video
		cursor *int64
	)
	for {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		page, err := c.ListVideos(ctx, ListVideosParams{
			Fields:   p.Fields,
			Cursor:   cursor,
			MaxCount: maxCount,
		})
		if err != nil {
			return nil, err
		}
		for i := range page.Videos {
			all = append(all, page.Videos[i])
			if p.Limit > 0 && len(all) >= p.Limit {
				return all, nil
			}
		}
		if !page.HasMore {
			return all, nil
		}
		next := page.Cursor
		cursor = &next
	}
}

// GetVideo fetches metadata for a single video by ID. Convenience wrapper
// around /v2/video/query/ for the common single-video case.
//
// Scope: video.list.
func (c *Client) GetVideo(ctx context.Context, videoID string, fields []string) (*Video, error) {
	if videoID == "" {
		return nil, errors.New("tiktok: GetVideo requires a non-empty video ID")
	}
	out, err := c.QueryVideos(ctx, []string{videoID}, fields)
	if err != nil {
		return nil, err
	}
	if len(out.Videos) == 0 {
		return nil, &Error{Code: "resource_not_found", Message: "video not found", StatusCode: http.StatusNotFound}
	}
	return &out.Videos[0], nil
}

// QueryVideos fetches metadata for a batch of videos by ID (1-20).
//
// Scope: video.list.
func (c *Client) QueryVideos(ctx context.Context, videoIDs []string, fields []string) (*VideoQueryResult, error) {
	if len(videoIDs) == 0 {
		return nil, errors.New("tiktok: QueryVideos requires at least one video ID")
	}
	if len(videoIDs) > 20 {
		return nil, fmt.Errorf("tiktok: QueryVideos accepts at most 20 IDs (%d provided)", len(videoIDs))
	}
	if len(fields) == 0 {
		return nil, errors.New("tiktok: QueryVideos requires at least one field")
	}

	q := url.Values{}
	q.Set("fields", strings.Join(fields, ","))

	body := map[string]any{
		"filters": map[string]any{"video_ids": videoIDs},
	}
	data, err := c.doJSON(ctx, http.MethodPost, c.baseURL+"/v2/video/query/", body, q)
	if err != nil {
		return nil, err
	}
	out := &VideoQueryResult{}
	if err := json.Unmarshal(data, out); err != nil {
		return nil, fmt.Errorf("tiktok: decode video query: %w", err)
	}
	return out, nil
}
